test/provenance.cpp

```diff
@@ -0,0 +1,297 @@
+    //         |   |
+    //        C{tag_c}
+    //
+    // Replacement:
+    //
+    //   C -> D{}
+    //
+    // After:
+    //
+    //   D{tag_a,tag_b,tag_c}
+    //
+    // Comment:
+    //   * D is the replacement root, and its insertion kills A, B, and C.
+    //
+    {
+        auto x = make_shared<op::Parameter>(element::i32, PartialShape{2, 3, 4});
+        auto y = make_shared<op::Parameter>(element::i32, PartialShape{2, 3, 4});
+
+        auto a = make_shared<op::Add>(x, y);
+        a->add_provenance_tag("tag_a");
+        auto b = make_shared<op::Multiply>(y, x);
+        b->add_provenance_tag("tag_b");
+        auto c = make_shared<op::Subtract>(a, b);
+        c->add_provenance_tag("tag_c");
+
+        auto f = make_shared<Function>(c, ParameterVector{x, y});
+
+        auto d = make_zero(element::i32, Shape{2, 3, 4});
+        replace_node(c, d, provenance_config);
+
+        EXPECT_EQ(d->get_provenance_tags(), (ProvSet{"tag_a", "tag_b", "tag_c"}));
+    }
+
+    //
+    // Before:
+    //
+    //   A{tag_a}  B{tag_b}
+    //         |   |
+    //        C{tag_c}
+    //
+    // Replacement:
+    //
+    //        D{}
+    //         |
+    //   C -> E{}
+    //
+    // After:
+    //
+    //   D{tag_a,tag_b,tag_c}
+    //   |
+    //   E{tag_a,tag_b,tag_c}
+    //
+    // Comment:
+    //   * E is the replacement root, and its insertion kills A, B, and C.
+    //   * D is post-dominated by E, so the tags inherited by E should also be taken on by D.
+    //
+    {
+        auto x = make_shared<op::Parameter>(element::i32, PartialShape{2, 3, 4});
+        auto y = make_shared<op::Parameter>(element::i32, PartialShape{2, 3, 4});
+
+        auto a = make_shared<op::Add>(x, y);
+        a->add_provenance_tag("tag_a");
+        auto b = make_shared<op::Multiply>(y, x);
+        b->add_provenance_tag("tag_b");
+        auto c = make_shared<op::Subtract>(a, b);
+        c->add_provenance_tag("tag_c");
+
+        auto f = make_shared<Function>(c, ParameterVector{x, y});
+
+        auto d = make_zero(element::i32, Shape{2, 3, 4});
+        auto e = make_shared<op::Negative>(d);
+        replace_node(c, e, provenance_config);
+
+        EXPECT_EQ(d->get_provenance_tags(), (ProvSet{"tag_a", "tag_b", "tag_c"}));
+        EXPECT_EQ(e->get_provenance_tags(), (ProvSet{"tag_a", "tag_b", "tag_c"}));
+    }
+
+    //
+    // Before:
+    //
+    //   A{tag_a}  B{tag_b}
+    //         |   |
+    //        C{tag_c}
+    //
+    // Replacement: C is replaced with G, where:
+    //
+    //     D{}  E{}
+    //      \  / \
+    //       G{}  F{}
+    //
+    // After:
+    //
+    //     D{tag_a,tag_b,tag_c}   E{}
+    //      \                    / \
+    //       G{tag_a,tag_b,tag_c}   F{}
+    //
+    // Comment:
+    //   * G is the replacement root, and its insertion kills A, B, and C.
+    //   * D is post-dominated by G, but E and F are not. Therefore D should take on the subsumed
+    //     tags, but E and F should not.
+    //
+
+    /*
```

Comment:
Why is this part commented out? A comment saying why would be good, or remove it.

Reply:
this is a test we intend to enable with a separate commit; it can be removed for this PR